### PR TITLE
JS validation: dynamically added form inputs are now validated

### PR DESCRIPTION
**Pull request by boris-brtan**
| Q             | A
| ------------- | ---
|Description, reason for the PR| problem with saving forms with dynamically added parameters
|New feature| No 
|BC breaks| No 
|Fixes issues| #190
|Standards and tests pass| Yes
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

### CODE REVIEW

File: packages/framework/src/Resources/scripts/common/validation/customizeBundle.js [hunk 1]
```diff
@@ -115,6 +115,22 @@
         Shopsys.timeout.setTimeoutAndClearPrevious('Shopsys.validation.validateWithParentsDelayed', executeDelayedValidators, 100);
     };
 
+    // Issue in dynamic collections validation
+    // https://github.com/formapro/JsFormValidatorBundle/issues/139
+    FpJsFormValidator._preparePrototype = FpJsFormValidator.preparePrototype;
+    FpJsFormValidator.preparePrototype = function (prototype, name, id) {
+        prototype.name = prototype.name.replace(/__name__/g, name);
```
Comment (vitek-rostislav) [inline packages/framework/src/Resources/scripts/common/validation/customizeBundle.js]: Could you please add some explanation what exactly happens here, i.e. what was the problem and how it was fixed? I see that you replace both prototype.name and prototype.id with given name but I do not understand why :slightly_frowning_face:

File: project-base/tests/ShopBundle/Acceptance/acceptance/PageObject/Admin/ProductEditPage.php [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+namespace Tests\ShopBundle\Acceptance\acceptance\PageObject\Admin;
+
+use Tests\ShopBundle\Acceptance\acceptance\PageObject\AbstractPage;
+
+class ProductEditPage extends AbstractPage
+{
+    /**
+     * @param int $productId
+     */
+    public function saveInvalidForm($productId)
```
Comment (vitek-rostislav) [inline project-base/tests/ShopBundle/Acceptance/acceptance/PageObject/Admin/ProductEditPage.php]: I would name this method more specifically - it saves form with invalid parameter...

File: project-base/tests/ShopBundle/Acceptance/acceptance/AdminProductSaveFormWithEmptyParametersCest.php [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        LoginPage $loginPage,
+        ProductEditPage $productEditPage
+    ) {
+        $me->wantTo('not able to save invalid form');
+        $loginPage->loginAsAdmin();
+
+        $productEditPage->saveInvalidForm(5);
+        $productEditPage->assertSaveFormErrorBoxVisible();
+
+        $productEditPage->cleanUpWorkspaceAfterChanges();
```
Comment (vitek-rostislav) [inline project-base/tests/ShopBundle/Acceptance/acceptance/AdminProductSaveFormWithEmptyParametersCest.php]: why? could not you just call `$this->tester->acceptPopup()`?

File: project-base/tests/ShopBundle/Acceptance/acceptance/AdminProductSaveFormWithEmptyParametersCest.php [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+use Tests\ShopBundle\Test\Codeception\AcceptanceTester;
+
+class AdminProductSaveFormWithEmptyParametersCest
+{
+    public function testSaveForm(
+        AcceptanceTester $me,
+        LoginPage $loginPage,
+        ProductEditPage $productEditPage
+    ) {
+        $me->wantTo('not able to save invalid form');
```
Comment (vitek-rostislav) [inline project-base/tests/ShopBundle/Acceptance/acceptance/AdminProductSaveFormWithEmptyParametersCest.php]: `$me->wantTo('see that I am not able to save an invalid form');`?

File: project-base/tests/ShopBundle/Acceptance/acceptance/PageObject/Admin/ProductEditPage.php [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    {
+        $this->tester->amOnPage('/admin/product/edit/' . $productId);
+        // scroll to have the button visible because of the fixed bars
```
Comment (vitek-rostislav) [inline project-base/tests/ShopBundle/Acceptance/acceptance/PageObject/Admin/ProductEditPage.php]: is it necessary to scroll down? I thought that when you do just `$this->tester->clickByCss('.js-parameters-item-add');`, it does not matter that the element is not in viewport, or does it?

File: CHANGELOG.md [hunk 6]
```diff
@@ -61,6 +61,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - [#260 - Admin: Js validation: dynamically added parameters are now validated](https://github.com/shopsys/shopsys/pull/260)
     - validation is evaluated after dynamic parameters are added
     - jQuery migration console error logging is no more
+    - support for multiple option of Flag and Brand items were added for form to be validated and not submitted if it is invalid
```
Comment (vitek-rostislav) [inline CHANGELOG.md]: I honestly do not know what to do with this problem, the implementation in `JsFormValidatorFactory::normalizeViewTransformers()` is (already was) really confusing and fragile :disappointed: It should probably be refactored completely but I do not know how... @PetrHeinz could you please help us here? :angel:

**Comment by vitek-rostislav** [comment]
While testing, I noticed yet another bug but it seems that it is already on master so I created a separate issue - https://github.com/shopsys/shopsys/issues/268

File: project-base/tests/ShopBundle/Acceptance/acceptance/PageObject/Admin/ProductEditPage.php (hunk 5 above)
Comment (boris-brtan) [reply project-base/tests/ShopBundle/Acceptance/acceptance/PageObject/Admin/ProductEditPage.php]: Yes, there is the problem with fixed elements. Click method will do the scroll to element like function ```javascript
document.querySelector('.js-parameters-item-add').scrollIntoView()
```
and then client will try to click on it but there is element fixed bar above it and codeception will fail on error 
```
Element \<a href="#" class="js-parameters-item-add btn btn--plus">...\</a> is not clickable at point (338, 15). Other element would receive the click: \<header class="web__header">...\</header>
```

File: project-base/tests/ShopBundle/Acceptance/acceptance/AdminProductSaveFormWithEmptyParametersCest.php (hunk 3 above)
Comment (boris-brtan) [reply project-base/tests/ShopBundle/Acceptance/acceptance/AdminProductSaveFormWithEmptyParametersCest.php]: In this test the form is changed and browser will stop execution of next tests due to modal box that wants approval. Seems that opposite order of commands doesn't work if the test is ended and next test requests for different page modal box is not canceled.

File: packages/framework/src/Resources/scripts/admin/parameters.js [hunk 7]
```diff
@@ -37,6 +37,7 @@
             Shopsys.formChangeInfo.showInfo();
             Shopsys.parameters.refreshCount($collection);
             Shopsys.validation.addNewItemToCollection('#product_edit_form_parameters', index);
+            $('#product_edit_form_parameters .js-parameters-item input').blur();
```
[the remark this answers is not in the document]
Comment (boris-brtan) [reply packages/framework/src/Resources/scripts/admin/parameters.js]: thx, yes better from UX view will be to focus into input

**Comment by PetrHeinz** [comment]
After discussion with @boris-brtan we've reduced scope of this PR as there are two unrelated issues atm:
- The dynamically added inputs do not have JS validation. This is what we aim to fix.
- On product edit page, the form is submitted even if the values are not valid, i.e. JS validation does not prevent the form from submitting. This is related to BG-2136 fixed in 3771de6, we will create a separate issue for this. Boris wanted to fix it in one of the last commits, which recieved most comment during this CR.

Can someone please do a code review of this fix with the reduced scope?

File: CHANGELOG.md [hunk 8]
```diff
@@ -78,6 +78,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - [#352 - flushes executed in loops are now executed outside of loop](https://github.com/shopsys/shopsys/pull/352)
 - [#342 - procedures are now executed only if relevant columns are changed](https://github.com/shopsys/shopsys/pull/342)
 - [#362 - guidelines-for-pull-request.md: fixed indentation of lines and code blocks](https://github.com/shopsys/shopsys/pull/362)
+- [#260 - JS validation: dynamically added form inputs are now validated](https://github.com/shopsys/shopsys/pull/260)
```
Comment (LukasHeinz) [inline CHANGELOG.md]: Please correct changelog entry because new version was released.

**Pull request merged**